File: lab10/src/by/bntu/fitr/minigame/domain/east/Vampire.java
package by.bntu.fitr.minigame.domain.east;

import by.bntu.fitr.minigame.domain.Creature;
import by.bntu.fitr.minigame.domain.Role;
import by.bntu.fitr.minigame.domain.Weapon;

public class Vampire extends Creature {
    private Weapon weapon = Weapon.DAGGER;
    private Role role = Role.THIEF;


    public Vampire(int health, int strength, int agility, int intelligence) {
        super(health, strength, agility, intelligence);
    }

    @Override
    protected int getWeaponDamage() {
        return weapon.getDamage();
    }

    @Override
    public int buff() {
        return (weapon.getDamage() + role.getPercent()) / 100 * getAgility() ;
    }
}

File: lab10kt/src/main/kotlin/by/bntu/fitr/povt/korotkevich/model/east/Vampire.kt
package by.bntu.fitr.povt.korotkevich.model.east

import by.bntu.fitr.povt.korotkevich.model.Creature
import by.bntu.fitr.povt.korotkevich.model.Role
import by.bntu.fitr.povt.korotkevich.model.Weapon
import by.bntu.fitr.povt.korotkevich.model.buff

class Vampire(override val health: Int = 100, override val strength: Int = 20,
              override val agility: Int = 50, override val intelligence: Int = 10,
              val role: Role = Role.THIEF, val weapon: Weapon = Weapon.DAGGER) : Creature() {

    override val id: String = "Vampire"

    override fun getSelfDamage() = weapon.damage + buff(role)

    override fun toString() = StringBuilder("Vampire(").apply {
        append(super.toString())
        append("role: $role, weapon: $weapon)")
    }.toString()
}